server/mailet/dkim/src/test/java/org/apache/james/jdkim/mailets/ConvertTo7BitTest.java

```diff
@@ -0,0 +1,229 @@
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jdkim.mailets;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+
+import javax.mail.BodyPart;
+import javax.mail.internet.MimeMultipart;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.james.util.MimeMessageUtil;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.FakeMailetConfig;
+import org.assertj.core.api.SoftAssertions;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import com.github.fge.lambdas.Throwing;
+
+class ConvertTo7BitTest {
+
+    private static final String CONTENT_TRANSFER_ENCODING = "Content-Transfer-Encoding";
+    private static final String X_MIME_AUTOCONVERTED = "X-MIME-Autoconverted";
+    private static final String QUOTED_PRINTABLE = "quoted-printable";
+    private static final String MESSAGE_BODY_8BIT = "A 8bit encoded body with €uro symbol.";
+    private static final String MESSAGE_BODY_QUOTED_PRINTABLE = "A 8bit encoded body with =E2=82=ACuro symbol.";
+    private static final String BASE64 = "base64";
+
+    private ConvertTo7Bit testee;
+
+    @BeforeEach
+    void setUp() throws Exception {
+        testee = new ConvertTo7Bit();
+        testee.init(FakeMailetConfig.builder()
+            .build());
+    }
+
+    @Nested
+    class When7BitMail {
+        @Nested
+        class WhenTextContent {
+            @Test
+            void serviceShouldKeepMessageContentUnTouch() throws Exception {
+                Mail mail = FakeMail.builder()
+                    .name("a-mail-with-quoted-printable-encoding")
+                    .mimeMessage(MimeMessageUtil.mimeMessageFromString(
+                        fileContent("eml/text-only-7bit.eml")))
```

Comment:
Should not the fileContent helper method method accept an encoding parameter, and we should call it using USASCII when opening 7bit files ?

Reply:
Very interesting question, it twisted my brain so badly for finding an answer. 
Firstly, if I make `fileContent` returning an InputStream directly without String representation, tests all still pass.

Secondly, ASCII is 7bit encoding, but the minimal data unit we are using to transfer data remotely or reading/writing with Java or even by the machines is byte(8bit). I guess there is a fixed bit padded in the chunk of 8bit which can be 0 or 1, then after that is 7bit data ASCII. 

So I can only read/write files(eml) with the chunk of 8bit(UTF-8). Thus, email data is actually formatted in UTF-8. But with a case of quoted-printable (8bit representation in 7bit format) email, you will see every character is ASCII. And a group of ASCII with 3 characters starting with =, like =3D actually represents for an 8bit character. 

The antique SMTP RFC [RFC 821](https://www.ietf.org/rfc/rfc0821.txt) mentions only ASCII data, that's means 8bit data is not taken into account. So I guess to follow the RFC, by the mean time, people still want to transfer non English characters by emails. They invented quoted-printable. After that, for how long IDK, 8BITMIME extension are out, then transferring 8bit characters are much more simpler.

Reply:
You are globally right. 7-bits means 8 bit characters where the highest bit is 0. So it means a char value of less than 127. It's interesting because it's the real ascii standard, or a subset of extended ascii that is encoded the same way in ascii, iso-xxx, utf-8, and probably many others.

Reply:
To answer the original question of @remk : we want to read arbitrary characters from the file, encoding the file as UTF-8 is probably a good solution.

ConvertTo7Bit then handles non-ascii characters as sequences of ascii characters (much like UTF-8, in fact).